src/darcy.erl

```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+
+do_scan_all(_Client, _Req, {ok, empty_table}, _Fun, _Acc) -> {ok, []};
+do_scan_all(_Client, _Req, {ok, no_results}, _Fun, Acc) -> {ok, lists:reverse(lists:flatten(Acc))};
+do_scan_all(_Client, _Req, {ok, #{ <<"Items">> := I }}, Fun, Acc) ->
+    {ok, lists:reverse(lists:flatten([ lists:map(Fun, I) | Acc ]))};
+do_scan_all(_Client, Req, {error, Error}, _Fun, Acc) ->
+    error_logger:error_msg("Error executing scan_all. request: ~p, error: ~p", [Req, Error]),
+    {error, {Error, Acc}};
+do_scan_all(Client, Req, {partial, #{ <<"LastEvaluatedKey">> := LEK, <<"Items">> := I }}, Fun, Acc) ->
+    NewRequest = make_scan_request([Req, #{ <<"ExclusiveStartKey">> => LEK }]),
+    do_scan_all(Client, NewRequest, execute_scan(Client, NewRequest), Fun, [ lists:map(Fun, I) | Acc]).
+
+%% @doc Scan a table in parallel using the given expression. This
+%% function is equivalent to `scan/7' with a timeout value of 60000
+%% milliseconds.
+%%
+%% Instead of returning the coordinator pid to the caller, this
+%% function blocks and waits for the return values from the
+%% coordinator.
+scan_parallel(Client, TableName, Expr, Fun, SegmentCount) ->
+    scan_parallel(Client, TableName, Expr, Fun, ?BIG_TIMEOUT, SegmentCount).
+
+scan_parallel(Client, TableName, Expr, Fun, Timeout, SegmentCount) ->
+    _ = scan_parallel(Client, TableName, Expr, Fun, Timeout, SegmentCount, self()),
+
+    receive
+        {Results, []} -> {ok, Results};
+        {Partial, Errors} -> {error, {Errors, Partial}};
+        Other -> Other
+    after Timeout ->
+        {error, scan_timeout}
+    end.
+
+%% @doc Scan a table in parallel using the given expression.
+%%
+%% DynamoDB supports parallel scans using a partitioning technique described in the Developer Guide.
+%%
+%% https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html
+%%
+%% This function spawns a coordinator process which in turn spawns a set of
+%% workers, one per segment count which executes the `scan_all/4' function.
+%%
+%% <B>N.B.</B>: This operation can consume a lot of read capacity. It is a good
+%% idea to limit the number of segments used in a scan operation.
+%%
+%% The coordinator that is spawned here is <B>linked</B> to the caller. If you
+%% want more robust error handling, you should trap exit messages by the
+%% caller's process.
+%%
+%% Since this function returns the "raw" results, you will have to
+%% handle them appropriately within your own receive block.
+-spec scan_parallel(
+        Client :: darcy_request:aws_client(),
+        TableName :: binary(),
+        Expr :: map(),
+        Fun :: function(),
+        Timeout :: pos_integer(),
+        SegmentCount :: pos_integer(),
+        ReplyPid :: pid()) -> scan_timeout |
+                              { Results :: [ term() ], Errors :: [ term() ] }.
+scan_parallel(Client, TableName, Expr, Fun, _Timeout, SegmentCount, ReplyPid) ->
+    Reqs = [ make_scan_request([Expr, table_name(TableName), make_segments(N, SegmentCount)]) ||
```

Comment:
Same comment as above, might be neat if you included a general 'parallel map with up to N workers' function and used that (?)